src/libcrun/kontain.c

```diff
@@ -0,0 +1,118 @@
+/*
+ * crun - OCI runtime written in C
+ *
+ * Copyright (C) 2020-2021 Kontain Inc.
+ * crun is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * crun is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with crun.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <config.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <argp.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/sysmacros.h>
+
+#include "kontain.h"
+#include "container.h"
+#include "utils.h"
+
+/*
+ * If execpath is not a symlink to km or the path to km, then assume we
+ * are to run execpath as a km payload.  In this case add km's path to argv[0]
+ * and shift argv[0 ... n] up to argv[1 ... (n+1)] and return the absolute
+ * path to km.
+ * Arguments:
+ *   argv[] is terminated with a null pointer.
+ *   execpath - the program the caller wants to run, absolute path
+ * Returns:
+ *   0 - success
+ *   errno - failure
+ */
+int
+libcrun_kontain_argv (char ***argv, const char **execpath)
+{
+  struct stat statb;
+  char **newargv;
+  const char *cmd = *execpath;
+
+  if (cmd[0] != '/')
+    { // verify that we are getting an absolute path
+      return EINVAL;
+    }
+  if (fstatat (AT_FDCWD, cmd, &statb, AT_SYMLINK_NOFOLLOW) != 0)
+    {
+      // the command does not exist?  Let the caller handle that.
+      return errno;
+    }
+  if (strcmp (cmd, KM_BIN_PATH) != 0)
+    {
+      /* Some program other than km, assume they need km to run it so insert the path to km into argv[] */
```

Comment:
what is he purpose of this?

Reply:
We want all programs that krun starts to run with km provided virtualization.
If a container entrypoint is already km then there is no need to add km to the arg vector as the program to run. 
If the entry point is not km we add it to the arg vector.